File: app/src/main/java/com/example/safeher/utils/MultiTapDetector.java
package com.example.safeher.utils;

public class MultiTapDetector {

    private final int requiredTaps;
    private final long windowMillis;
    private int currentTapCount;
    private long firstTapTimestamp;

    public MultiTapDetector(int requiredTaps, long windowMillis) {
        this.requiredTaps = requiredTaps;
        this.windowMillis = windowMillis;
        this.currentTapCount = 0;
        this.firstTapTimestamp = 0L;
    }

    public boolean registerTap() {
        long now = System.currentTimeMillis();

        if (firstTapTimestamp == 0L || (now - firstTapTimestamp > windowMillis)) {
            firstTapTimestamp = now;
            currentTapCount = 1;
            return false;
        }

        currentTapCount++;
        if (currentTapCount >= requiredTaps) {
            reset();
            return true;
        }

        return false;
    }

    public void reset() {
        firstTapTimestamp = 0L;
        currentTapCount = 0;
    }
}

File: ios/MultiTapDetector.swift
//
//  MultiTapDetector.swift
//  Protecta
//
//  Created by GitHub Copilot
//

import Foundation

class MultiTapDetector {
    private let requiredTaps: Int
    private let timeWindow: TimeInterval
    private var currentTapCount = 0
    private var firstTapTimestamp: TimeInterval = 0

    init(tapsRequired: Int, timeWindow: TimeInterval) {
        self.requiredTaps = tapsRequired
        self.timeWindow = timeWindow
    }

    func registerTap() -> Bool {
        let now = Date().timeIntervalSince1970

        if firstTapTimestamp == 0 || (now - firstTapTimestamp > timeWindow) {
            firstTapTimestamp = now
            currentTapCount = 1
            return false
        }

        currentTapCount += 1
        if currentTapCount >= requiredTaps {
            reset()
            return true
        }

        return false
    }

    func reset() {
        firstTapTimestamp = 0
        currentTapCount = 0
    }
}